hls-graph/src/Development/IDE/Graph/Internal/Database.hs

```diff
@@ -0,0 +1,144 @@
+import qualified Development.Shake.Rule as Shake
+import Control.Exception
+import Control.Monad.Trans.Reader
+import Data.Tuple.Extra
+import Data.Either
+
+newDatabase :: Dynamic -> TheRules -> IO Database
+newDatabase databaseExtra databaseRules = do
+    databaseStep <- newIORef $ Step 0
+    databaseLock <- newLock
+    databaseIds <- newIORef Intern.empty
+    databaseValues <- Ids.empty
+    pure Database{..}
+
+incDatabase :: Database -> IO ()
+incDatabase db = do
+    modifyIORef' (databaseStep db) $ \(Step i) -> Step $ i + 1
+    Ids.forMutate (databaseValues db) $ second $ \case
+        Clean x -> Dirty (Just x)
+        Dirty x -> Dirty x
+        Running _ x -> Dirty x
+
+
+build
+    :: forall key value . (Shake.RuleResult key ~ value, Typeable key, Show key, Hashable key, Eq key, Typeable value)
+    => Database -> [key] -> IO ([Id], [value])
+build db keys = do
+    (ids, vs) <- fmap unzip $ builder db $ map (Right . Key) keys
+    pure (ids, map (asV . resultValue) vs)
+    where
+        asV :: Value -> value
+        asV (Value x) = unwrapDynamic x
+
+builder
+    :: Database -> [Either Id Key] -> IO [(Id, Result)]
+builder db@Database{..} keys = do
+    -- Async things that I own and am responsible for killing
+    ownedAsync <- newIORef []
+    flip onException (cleanupAsync ownedAsync) $ do
+
+        -- Things that I need to force before my results are ready
+        toForce <- newIORef []
+
+        results <- withLock databaseLock $ do
+            forM keys $ \idKey -> do
+                id <- case idKey of
+                    Left id -> pure id
+                    Right key -> do
+                        ids <- readIORef databaseIds
+                        case Intern.lookup key ids of
+                            Just v -> pure v
+                            Nothing -> do
+                                (ids, id) <- pure $ Intern.add key ids
+                                writeIORef' databaseIds ids
+                                return id
+
+                status <- Ids.lookup databaseValues id
+                val <- case fromMaybe (fromRight undefined idKey, Dirty Nothing) status of
+                    (_, Clean r) -> pure r
+                    (_, Running act _) -> do
+                        -- we promise to force everything in todo before reading the results
+                        -- so the following unsafePerformIO isn't actually unsafe
+                        let (force, val) = splitIO act
+                        modifyIORef toForce (force:)
+                        pure val
+                    (key, Dirty s) -> do
+                        -- Important we don't lose any Async things we create
+                        act <- uninterruptibleMask $ \restore -> do
+                            -- the child actions should always be spawned unmasked
+                            -- or they can't be killed
+                            async <- async $ restore $ check db key id s
```

Comment:
On IRC, we were discussing if it was possible to make all the book-keeping contention free, by eliminating the locks and running in STM (and using a concurrent map like [stm-containers](https://hackage.haskell.org/package/stm-containers-1.2)?).

This bit seems to be the problematic part, the rest of the computation seem like it can easily be moved into STM. But it might just be possible to make something work using judicious use of `unsafePerformIO`, `unsafeIOToSTM` and the proper care. Do you think this would be possible?